### PR TITLE
 Merging django-ipam & openwisp-ipam

### LINKED ISSUE #2: [ipam] Improve the README file.
Improve the README file by taking inspiration from the `django-ipam` README.

### PR DESCRIPTION
**Pull request by atb00ker**
Closes #2 as well.

### CODE REVIEW

**Comment by atb00ker** [comment]
-- To be noted: The `sample_ipam` app is testing `ipam` without openwisp-users as I assume some users would want to use `ipam` without multitenancy and it may serve as an example for them as well! :smile:

**Comment by atb00ker** [comment]
Hi, 

I understand the part of adding organization in base model, however, 
I didn't get : 
> I noticed the organization field moved down in the admin interface.
> Which is something we don't want.

@nemesisdesign The organization field looks like exactly where it was before, I think I am misunderstanding something here.

**Comment by atb00ker** [comment]
> If anyone wants to use this module without multi-tenancy, they can do so by removing the organization field in their custom models.

Made the change! :smile:

File: openwisp_ipam/tests/test_users_integration.py [hunk 1]
```diff
@@ -1,11 +0,0 @@
-from openwisp_users.tests.test_admin import TestUsersAdmin
```
Comment (nemesisdesign) [inline openwisp_ipam/tests/test_users_integration.py]: why is this removed?

Comment (atb00ker) [reply openwisp_ipam/tests/test_users_integration.py]: Commented it out during testing, forgot to add it back, thanks for pointing it out.
Updated.

**Comment by atb00ker** [comment]
> Another PR merged in django-ipam, minimal changes: openwisp/django-ipam#106

**Updated** :smile:

**Comment by atb00ker** [comment]
> Another PR merged in django-ipam, minimal changes: openwisp/django-ipam#106

**Updated** :smile:

File: README.rst [hunk 2]
```diff
@@ -44,11 +44,476 @@ Available Features
...
+    <ip-address>,<optional-description>
+    <ip-address>,<optional-description>
+
+------------
+
+Extending openwisp-ipam
+=====================
+
+Extending API Views
+###################
```
Comment (nemesisdesign) [inline README.rst:460]: Don't we need instructions also on how to extend/reuse other parts?

- Models
- Admn
- AppConfig
- Tests

I think it should also be possible to provide a swapped model without having to create a different django app.
Would be nice to explain that as well and provide a working example.
We can reuse the same template for the other projects.

**Comment by atb00ker** [comment]
1. I added `EXTEND.rst` for documenting all the steps for extending the application. This file can be reused in other modules with small changes.

> I think it should also be possible to provide a swapped model without having to create a different django app.

@nemesisdesign  Sorry, I didn't get it, do you mean that the user wants to only change the models. (i.e Create `models.py` in their project and use swapper's settings `OPENWISP_IPAM_IPADDRESS_MODEL` and `OPENWISP_IPAM_SUBNET_MODEL` to swap the modules? Not sure If this is what you mean.)

**Comment by atb00ker** [comment]
> Yes that's what I meant.
> That should be possible if I'm not mistaken.

If I only declare `models.py` & use the settings, I wouldn't get the new models registered in admin for which I'll need to create `admin.py` file and to run tests on it, I'll have to declare `tests.py` and copy management commands from `openwisp_ipam`.

So, unless I am thinking something wrong / I am unaware of a feature, I don't think It would be possible to just change `models.py`.

File: EXTEND.rst [hunk 3]
```diff
@@ -0,0 +1,176 @@
+Extending openwisp-ipam
```
Comment (nemesisdesign) [inline EXTEND.rst:1]: Ajay let's not divide the README.
In the future, when we'll improve the docs we may put effort in dividing the content in different page, but now let's just offer everything in one document for consistency with all the other modules (it's also better from an SEO standpoint in this case, as the first page of each repo has better ranking).

File: .travis.yml [hunk 4]
```diff
@@ -11,26 +11,24 @@ python:
...
-  - pip install -e git+https://github.com/openwisp/django-ipam#egg=django-ipam
   - pip install --no-cache-dir -U -r requirements-test.txt
 
 install:
   - pip install $DJANGO
   - python setup.py -q develop
 
-# command to run tests, e.g. python setup.py test
 script:
+  - ./run-qa-check.sh
```
Comment (nemesisdesign) [inline .travis.yml:29]: should be `run-qa-checks.sh` (plural)

File: EXTEND.rst [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    class ImportSubnetView(BaseImportSubnetView):
+        subnet_model = Subnet
+        queryset = Subnet.objects.none()
+
+
+    class ExportSubnetView(BaseExportSubnetView):
+        subnet_model = Subnet
+        queryset = Subnet.objects.none()
+
+
```
Comment (nemesisdesign) [inline EXTEND.rst:169]: do not leave too many blank lines please

File: EXTEND.rst (hunk 3 above)
Comment (atb00ker) [reply EXTEND.rst:1]: Okay (I have to make that change it all the PRs)! :smile:

File: .travis.yml [hunk 6]
```diff
@@ -11,26 +11,24 @@ python:
 branches:
   only:
     - master
+    # TODO: Remove when merging to master
+    - gsoc-2020
```
Comment (nemesisdesign) [inline .travis.yml:15]: I think we can merge in master

File: openwisp_ipam/static/openwisp-ipam/js/subnet.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                pageDown();
+            } else if (height * 0.25 >= scrollTop) {
+                pageUp();
+            }
+        }
+    }
+    $('.subnet-visual').scroll(onUpdate);
+    onUpdate();
+}
+
```
Comment (nemesisdesign) [inline openwisp_ipam/static/openwisp-ipam/js/subnet.js:123]: is this last empty line needed? If not remove it please

File: README.rst [hunk 8]
```diff
@@ -44,11 +44,585 @@ Available Features
 Dependencies
 ============
 
-* Python 3.4 or higher
-* Django 2.0 or higher
```
Comment (nemesisdesign) [inline README.rst:48]: can we take advantage of this update to update the screenshots? For example, it would be nice to show the infinite scroller, if it doesn't take too much time, could you create a gif with the action of scrolling a huge ipv6 subnet?
I think it's a cool feature we should promote.

If you see any other outdated screenshot, or any other outdated part of the README which needs improvement, please fix it.

File: README.rst [hunk 9]
```diff
@@ -44,11 +44,585 @@ Available Features
 Dependencies
 ============
```
[the remark this answers is not in the document]
Comment (atb00ker) [reply README.rst:46]: \imo it's not really a "feature" of the application that concerns user when they are reading the README to understand what the project is accomplishing. I would consider infinite scroll as a UI component instead and a better way to show UI components is images / video / gifs! :smile:

But if you want to add, let me know! :smile:

[the remark this answers is not in the document]
Comment (nemesisdesign) [reply README.rst:46]: yes you're right, no need to mention the infinite scroller, we should just make sure it's clear that users can easily and quickly browse the address space.
I would still like to show the infinite scroller in a gif because I think it's a cool feature that may result useful to some.

File: openwisp_ipam/api/generics.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
+import csv
```
[the remark this answers is not in the document]
Comment (atb00ker) [reply openwisp_ipam/api/generics.py]: Done! :smile:

File: README.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (atb00ker) [reply README.rst:45]: Done! :smile:

File: README.rst (hunk 8 above)
Comment (atb00ker) [reply README.rst:48]: Okay, I added a GIF, do you prefer this GIF or a cropped version of it only focusing on infinite scroll? :smile:

Comment (nemesisdesign) [reply README.rst:48]: Will write in chat
